Fix tracing for modules with List[Tensor] as output

volkhin opened the pull request:
Summary: Fix an issue in TorchScript tracing for modules with `c10::List<at::Tensor>` as an output. TensorList was not supported properly.

Test Plan: unit test

Differential Revision: D18850722

kostmo commented:
## CircleCI build failures summary
As of commit 8583e0c:
* **1/1** broken upstream at merge base c5d3be1 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=c5d3be1102cc25c884eea472a5b28a0eb157d39a&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=8583e0cfcf91f8ad2ece4f30db3a5a12fcdc62cf).

### 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [caffe2_onnx_py2_gcc5_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=75507199)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

volkhin commented:
cc @suo @wanchaol

wanchaol commented on torch/csrc/jit/tracer.cpp:
```diff
@@ -200,6 +200,14 @@ Value* TracingState::getOutput(const IValue& iv, size_t i) {
        throw std::runtime_error(os.str());
      }
      return it->second;
+  } else if (iv.isTensorList()) {
```
you should also edit the error message to reflect the fact that list of tensor can be output of traced functions.

wanchaol commented on torch/csrc/jit/tracer.h:
```diff
@@ -308,20 +308,22 @@ TORCH_API void ensureUniqueIfOutOfPlaced(
     typename T,
     typename = torch::enable_if_t<
         (!std::is_convertible<torch::decay_t<T>, at::TensorList>::value &&
+         !std::is_convertible<torch::decay_t<T>, c10::List<at::Tensor>>::value &&
```
same, at::TensorList should be the same as c10::List<at::Tensor>

volkhin replied:
Right, it should work. I think the problem was that we provided implementation for std::vector<...> instead of c10::List<...>.

wanchaol commented on torch/csrc/jit/tracer.h:
```diff
@@ -308,20 +308,22 @@ TORCH_API void ensureUniqueIfOutOfPlaced(
+TORCH_API void addOutput(Node* node, const at::Tensor& tensor);
+TORCH_API void setOutput(Value* value, const at::Tensor& output);
+TORCH_API void addOutput(Node* node, const std::vector<at::Tensor>& list);
+TORCH_API void addOutput(Node* node, const c10::List<at::Tensor>& list);
```
std::vector<at::Tensor>& list and `c10::List<at::Tensor>& list` seems duplicate

volkhin replied:
Actually std::vector is unnecessary here, but without c10::List it will not work. I don't think there is any auto conversion between them.

wanchaol commented on torch/csrc/jit/tracer.cpp:
```diff
@@ -641,6 +649,18 @@ void addOutput(Node* node, const std::vector<at::Tensor>& outputs) {
   }
 }
 
+void addOutput(Node* node, const c10::List<at::Tensor>& outputs) {
```
there's a function `addOutput` that already takes `std::vector<at::Tensor>& outputs`, you should just use that. why we need to add a new case here?

volkhin replied:
they have slightly different interfaces from what I understand

wanchaol replied:
The list different between c10::List and std::vector is not ideal. @zdevito is working on removing the list specialization in https://github.com/pytorch/pytorch/pull/30734, which hopefully can allow the tracer takes `std::vector<T>` as a constructor of `c10::List`. In that case we should only have one `addOutput` here which takes c10::List<at::Tensor> only.

volkhin replied:
@wanchaol thanks for letting me know. What should we do here? Wait for #30734 to get merged or proceed as is for now?

wanchaol replied:
I think we can just merge, @zdevito will do a simple rebase fix after we merged.

volkhin commented on test/test_jit.py:
```diff
@@ -9949,6 +9949,14 @@ def foo(a):
 
+    def test_trace_chunk_with_list_output(self):
+        def chunk(x):
+            return torch.ops.aten.chunk(x, chunks=2, dim=0)
```
It doesn't repro with torch.chunk

wanchaol replied:
hmm ok it does not repro with `torch.chunk`, mainly because `torch.chunk` in tracing is handing in VariableType level where it produce `std::vector<at::Tensor>`. 

I think here we should not use the chunk or `torch.ops.aten.chunk` as a test case, because `torch.ops.aten.chunk` is JIT binded custom op that might get changed(although unlikely) in the future. 

What this PR does is basically enable List as the output of the tracer, what we should do here for testing is to add test cases that cover the below use cases:
1. List[Tensor] as trace output, should works and not throw
2. List[other types], should throw to user with the correct error message reporting.
3. nested lists, what should be valid in this case? We should only throw to user if at least one element type is not Tensor. 


Can you add at least these three test cases? For example for case one we should test if the below function works for tracing:

```python
def trace_list_output(x):
    return [x, x, x]
```
similarly for case 2 and 3.

Also, we could just use the `self.checkTrace` helper instead of manual tracing in the test cases. You can refer to https://github.com/pytorch/pytorch/blob/5579611544c6191d147fe4c7d7f5f5def6f44315/test/test_jit.py#L2223 to see how we add those test cases.

volkhin replied:
Good point! Adding tests for these cases. I'll not support nested lists for now, making sure it throws an exception.

wanchaol commented on torch/csrc/jit/tracer.cpp:
```diff
@@ -208,7 +216,7 @@ Value* TracingState::getOutput(const IValue& iv, size_t i) {
   } else {
     AT_ERROR(
-        "Only tensors or tuples of tensors can be output from traced functions");
+        "Only tensors and list/tuples of tensors can be output from traced functions");
```
nit: `Only tensors, lists or tuples of tensors can be output from traced functions`

wanchaol reviewed (changes requested):
Nice! we are getting close! I think we need to refactor the test cases to ensure we cover most use cases for List as an output of the tracer, then we are good to go.

volkhin reviewed (commented):
thanks for the review! Updating diff.

volkhin committed:
Fix tracing for modules with List[Tensor] as output (#31343)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31343

Fix an issue in TorchScript tracing for modules with `c10::List<at::Tensor>` as an output. TensorList was not supported properly.

Test Plan: unit tests

Differential Revision: D18850722

fbshipit-source-id: 8d438e69dfbd54b3bf393250073b4fb2945ad9f9

volkhin commented:
updated

wanchaol reviewed (approved):
looks good, thanks for adding the feature!
